### PR TITLE
[RISCV][GISel] Do libcall for G_FPTOSI, G_FPTOUI when no D or F support

### PR DESCRIPTION
**Pull request by spaits**
When compiling the following code:
```cpp
#include <stdio.h>
#include <stdlib.h>
#include <stddef.h>
#include <stdbool.h>

int main() {
    int a;
    float f;
    scanf("%d", &a);

    scanf("%f", &f);
    a += (int)f;
    
    return a;
}
``` 
for `-march=rv32ima_zbb` we get a libcall:
```
call    scanf
        lw      a0, -20(s0)
        call    __fixsfsi
        mv      a1, a0
```
When we try to use GlobalISel we get this error:
```
 error in backend: unable to legalize instruction: %9:_(s32) = G_FPTOSI %8:_(s32) (in function: main)
```

(Here is a link to a reproducer in Godblot: https://godbolt.org/z/f67vEEb41 )

The goal of this PR is to do a libcall for the legalization of `G_FPTOSI` and `G_FPTOUI` instead of doing a fallback to Selection DAG to do the same libcall later.

### CODE REVIEW

File: llvm/test/CodeGen/RISCV/GlobalISel/legalizer/legalize-fptoi-rv32-libcall.mir [hunk 1]
```diff
@@ -0,0 +1,194 @@
+# NOTE: Assertions have been autogenerated by utils/update_mir_test_checks.py
```
Comment (topperc) [inline llvm/test/CodeGen/RISCV/GlobalISel/legalizer/legalize-fptoi-rv32-libcall.mir:1]: We need an rv64 test too

File: llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp [hunk 2]
```diff
@@ -17,6 +17,7 @@
 #include "llvm/CodeGen/GlobalISel/GIMatchTableExecutor.h"
 #include "llvm/CodeGen/GlobalISel/GenericMachineInstrs.h"
 #include "llvm/CodeGen/GlobalISel/LegalizerHelper.h"
+#include "llvm/CodeGen/GlobalISel/LegalizerInfo.h"
```
Comment (topperc) [inline llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp:20]: Why do we need a new include?

File: llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp [hunk 3]
```diff
@@ -169,7 +170,8 @@ RISCVLegalizerInfo::RISCVLegalizerInfo(const RISCVSubtarget &ST)
       MergeUnmergeActions.legalIf(
           all(typeIs(BigTyIdx, s64), typeIs(LitTyIdx, s32)));
     }
-    MergeUnmergeActions.widenScalarToNextPow2(LitTyIdx, XLen)
+    MergeUnmergeActions.legalIf(all(typeIs(BigTyIdx, sDoubleXLen), typeIs(LitTyIdx, sXLen)))
```
Comment (topperc) [inline llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp:173]: How do we isel this UNMERGE?

File: llvm/test/CodeGen/RISCV/GlobalISel/legalizer/legalize-fptoi-rv32.mir [hunk 4]
```diff
@@ -166,13 +265,27 @@ body:             |
...
+    ; CHECK-FLOAT-NEXT: {{  $}}
+    ; CHECK-FLOAT-NEXT: [[COPY:%[0-9]+]]:_(s64) = COPY $f10_d
+    ; CHECK-FLOAT-NEXT: [[FPTOSI:%[0-9]+]]:_(s32) = G_FPTOSI [[COPY]](s64)
+    ; CHECK-FLOAT-NEXT: $x10 = COPY [[FPTOSI]](s32)
+    ; CHECK-FLOAT-NEXT: PseudoRET implicit $x10
+    ;
+    ; CHECK-NOFLOAT-LABEL: name: fptosi_s1_s64
+    ; CHECK-NOFLOAT: liveins: $f10_d
+    ; CHECK-NOFLOAT-NEXT: {{  $}}
+    ; CHECK-NOFLOAT-NEXT: [[COPY:%[0-9]+]]:_(s64) = COPY $f10_d
```
Comment (topperc) [inline llvm/test/CodeGen/RISCV/GlobalISel/legalizer/legalize-fptoi-rv32.mir:279]: Something is broken here. We're passing f10_d, but we don't have floating point?

**Comment by spaits** [comment]
@topperc I have updated the PR. I have discovered, that I could move the test for the libcall cases into the same file in which the non-libcall test are with the help of file check. When running the tests fail in some cases with this:
```
LLVM ERROR: unable to legalize instruction: %4:_(s32), %5:_(s32) = G_UNMERGE_VALUES %0:_(s64) (in function: fptosi_s1_s64)
```

So it says that we have an illegal unmerge that can not be legalized. This does not makes sense to me. I was editing the test for merge a few hours back today and I think this should be legalized by the current rules.

I have then realized, that when we are not on an rv32 system that has D extension we never have a `legalIf` or `legalFor` call. When I add a `legalFor` call then the legalization will be fine.

This is the commit, that introduced that if that made the `legalFor` call for only a specific case: https://github.com/llvm/llvm-project/commit/dbb9043dea238fa4e5b6a9a7fef99623b543493e .
Looking at the instruction selection and register bank info I don't really see why we shouldn't have a `legalFor` when the D extension is not present.

How should I proceed with this PR? Why is there no legal `G_UNMERGE` call when we have a CPU that is not rv32 and has no D extension?

I would really appreciate your help.

File: llvm/test/CodeGen/RISCV/GlobalISel/legalizer/legalize-fptoi-rv32.mir (hunk 4 above)
Comment (spaits) [reply llvm/test/CodeGen/RISCV/GlobalISel/legalizer/legalize-fptoi-rv32.mir:279]: So f10_d is a register only available when D and F are enabled. So I should be using for example x10 as a livein register right?

Comment (topperc) [reply llvm/test/CodeGen/RISCV/GlobalISel/legalizer/legalize-fptoi-rv32.mir:279]: For rv32, you would need to use x10 and x11 to hold all 64 bits. I recommend writing an IR test and using -stop-before=legalizer to get the MIR

**Review by topperc — APPROVED** [review]
LGTM

**Pull request merged**